proposal/shengwen-tw/README.rst

```diff
@@ -0,0 +1,142 @@
+#################################################################
+libqpsolver: a library for solving quadratic programming problems
+#################################################################
+
+Basic information
+#################
+
+* A numerical optimizer for solving quadratic cost functions with or without equality / inequality constraints
+
+* The optimizer will be implemented in C. (Not C++ since author will like to integrate it into embedded devices later.)
+
+* The future goal of this project is to integrate it in to embedded devices, which requires more improvments to
+  speed up the code. I expect myself to finish the first version this semester and able to run it on normal PC.
+
+GitHub: https://github.com/shengwen-tw/libqpsolver
+
+Problem to solve
+################
+
+What is Quadratic Programming?
+------------------------------
+
+.. image:: qp_formula.PNG
+
+We call the variable x **optimization variables** to be minimized,
+
+the first term **objective function** or **cost function**,
+
+the second term **equality constraints**,
+
+and the third term **inequality constraints**.
+
+Why do we need Quadratic Programming?
+-------------------------------------
+
+Quadratic Programming shows up frequently in robotics field like control or trajectory planning.
+
+* For example, if we want to map force and torque (6 degrees total) into motors numbers greater than 6,
+then we have infinite combinations for choosing motor thrust.
+
+* One of the solution is to choose the combination which has the lowest energy (in the sense of quadratic form)
+
+There are two fantasic examples to illustate this problem:
+
+.. image:: dragon.PNG
+
+link: https://www.youtube.com/watch?v=uje6iUBbkwM
+
+.. image:: omnicopter.PNG
+
+link: https://www.youtube.com/watch?v=sIi80LMLJSY&t=27s
+
+
+Perspective users
+#################
+
+I will expect the user may have the following backgrounds:
+
+1. Robotics
+
+2. Control
+
+Usually the problem set up will like: 
+
+We want robots to do something, but there are multiple choices.
+We then construct a quadratic cost function to analogize the energy cost and choose the least one strategy.
+
+System architecture
+###################
+
+.. image:: architecture.PNG
+
+API description
+###############
+
+we use **qp_t** datatype to store all informations that the solver need to solve the problem
```

Comment:
Do you plan to use the C-style API?  If yes, how do you plan to implement encapsulation?

Reply:
Yes, please check my early implementation (I'll add this in the proposal later also):
https://github.com/shengwen-tw/libqpsolver/blob/master/src/main.c#L32

By having data structure type like: **matrix_t**,
I'll create several interface function like:
``matrix_construct()``, ``matrix_new()``, ``matrix_zeros()`` to avoid user to do low level operations like malloc, calloc, etc.

Since C does not have much syntax sugars like C++, I may use macros to expand the language.
For example, I had define such code for matrix initialization:
```
#define ELEMENTS(...) (FLOAT []){__VA_ARGS__}

matrix_t P;
matrix_construct(&P, 2, 2, ELEMENTS(+1, -1, -1, +2));
```

and matrix element getter/setter:

```
#define matrix_at(mat_ptr, r, c) (mat_ptr)->data[(r * (mat_ptr)->column) + c]
```

Other C preprocessor techniques like ``stringification`` and ``string concatenation`` may be applied to help me to simply my code.

Reply:
Thanks.  I'll recommend to apply the C++ techniques that you will learn later in the course.  We can leave the choice of language here.  Being explicit in the proposal is good enough for now.